optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/common/BavetNodeBuildPolicy.java

```diff
@@ -26,7 +26,7 @@
 
     private final BavetConstraintSession session;
 
-    private int nodeOrderMaximum = 0;
+    private int nodeIndexMaximum = -1; // So that the first node starts at 0 when it increments.
```

Comment:
My suggestion is to remove this field and instead have

```suggestion
    private int nextNodeIndex = 0;
    
    public int nextNodeIndex() {
        return nextNodeIndex++;
    }
```

Every time a new node is created, it call this method and automatically gets the next one.
Remember that join nodes are created after both bridges are created, so a join will never be before it's bridges or anything before it (but we should test those nodeIndexes with a unit test).

Reply:
No. The entire algorithm depends on the fact, that `nodeIndex` is only ever actually incremented when node sharing is ruled out. It makes it incredibly simple and effective - no need for any sort of rollback.